### PR TITLE
Ports the weather system from Nebula.

### PR DESCRIPTION
**Pull request by NonQueueingMatt**
todo:

- [x] backport abstract markers from raccoon city
- [x] backport the umbrella stuff
- [x] test if it works on exoplanets

### CODE REVIEW

**Comment by NonQueueingMatt** [comment]
!review

File: code/datums/state_machine/state.dm [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// An individual state, defined as a `/decl` to save memory.
```
Comment (Geevies) [inline code/datums/state_machine/state.dm:1]:
```suggestion
// An individual state, defined as a `/singleton` to save memory.
```

**Review by Geevies — APPROVED** [review]

**Review by SleepyGemmy — APPROVED** [review]

File: code/game/objects/items.dm [hunk 2]
```diff
@@ -1227,3 +1227,6 @@ modules/mob/living/carbon/human/life.dm if you die, you will be zoomed out.
 
 /obj/item/proc/is_shovel()
 	return FALSE
+
+/obj/item/proc/gives_weather_protection()
+	return FALSE
```
Comment (Arrow768) [inline code/game/objects/items.dm:1232]: Imho this should include the weather it is protecting from as a parameter.
i.e. a umbrella might protect you during a "normal" rainshower, but when you get to a actual (snow-)storm the umbrella becomes useless.

File: code/modules/mob/living/life.dm [hunk 3]
```diff
@@ -189,3 +186,57 @@
...
+
+	SHOULD_CALL_PARENT(TRUE)
+
+	// Handle physical effects of weather.
+	var/singleton/state/weather/weather_state
+	var/obj/abstract/weather_system/weather = get_affecting_weather()
+	if(weather)
+		weather_state = weather.weather_system.current_state
+		if(istype(weather_state))
+			weather_state.handle_exposure(src, get_weather_exposure(weather), weather)
```
Comment (Arrow768) [inline code/modules/mob/living/life.dm:217]: You are already passing the weather as a parameter here

File: code/modules/mob/mob.dm [hunk 4]
```diff
@@ -1395,6 +1395,58 @@
...
+		var/obj/item/clothing/head/check_head = get_equipped_item(slot_head_str)
+		if(!istype(check_head) || !check_head.protects_against_weather)
+			return
+		var/obj/item/clothing/suit/check_body = get_equipped_item(slot_wear_suit_str)
+		if(!istype(check_body) || !check_body.protects_against_weather)
+			return
+		LAZYADD(., check_head)
+		LAZYADD(., check_body)
+
+/mob/proc/get_weather_exposure()
```
Comment (Arrow768) [inline code/modules/mob/mob.dm:1420]: But you dont evaluate the weather parameter you are passing here

**Review by Arrow768 — CHANGES_REQUESTED** [review]

**Pull request merged**